frame/treasury-oracle/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+//! * Conversion rates are only required from some asset to native.
+//!
+//! ## Related Modules
+//! * [`Treasury`](../treasury/index.html)
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::traits::{
+	fungible::Inspect,
+	tokens::{Balance, BalanceConversion},
+};
+use frame_system::WeightInfo;
+use sp_runtime::{traits::Zero, FixedPointNumber, FixedPointOperand, FixedU128};
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+// #[cfg(feature = "runtime-benchmarks")]
+// mod benchmarking;
+
+// Type alias for `frame_system`'s account id.
+type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+// This pallet's asset id and balance type.
+type AssetIdOf<T> = <T as Config>::AssetId;
+// Generic fungible balance type.
+type BalanceOf<T> = <<T as Config>::Currency as Inspect<AccountIdOf<T>>>::Balance;
+
+// #[frame_support::pallet]
+// TODO: Remove
+#[frame_support::pallet(dev_mode)]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+
+		/// The runtime event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The origin permissioned to create a conversion rate for an asset.
+		type CreateOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The origin permissioned to remove an existing conversion rate for an asset.
+		type RemoveOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The origin permissioned to update an existiing conversion rate for an asset.
+		type UpdateOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The units in which we record balances.
+		type Balance: Balance + FixedPointOperand;
+
+		/// The currency mechanism for this pallet.
+		type Currency: Inspect<Self::AccountId, Balance = Self::Balance>;
+
+		/// The identifier for the class of asset.
+		type AssetId: Member + Parameter + Copy + MaybeSerializeDeserialize + MaxEncodedLen;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn conversion_rate_to_native)]
+	/// Maps an asset to its fixed point representation in the native balance.
+	pub(super) type ConversionRateToNative<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssetId, FixedU128, OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		// Some `asset_id` conversion rate was created.
+		Created { asset_id: T::AssetId, rate: FixedU128 },
+		// Some `asset_id` conversion rate was removed.
+		Removed { asset_id: T::AssetId },
+		// Some existing `asset_id` conversion rate was updated from `old` to `new`.
+		Updated { asset_id: T::AssetId, old: FixedU128, new: FixedU128 },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The given asset ID is unknown.
+		Unknown,
```

Comment:
```suggestion
		UnknownAssetId,
```

without reading the doc, it means Unknown Error